test/jdk/sun/nio/ch/ChannelInputStream/TransferTo.java

```diff
@@ -0,0 +1,233 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import static java.lang.String.format;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.net.URLDecoder;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+
+/*
+ * @test
+ * @bug 8265891
+ * @summary tests whether sun.nio.ChannelInputStream.transferTo conforms to the
+ *          InputStream.transferTo contract defined in the javadoc
+ * @key randomness
+ */
+public class TransferTo {
+
+    public static void main(String[] args) throws Exception {
+        test(fileChannelInput(), writableByteChannelOutput());
+        test(seekableByteChannelInput(), fileChannelOutput());
+        test(readableByteChannelInput(), fileChannelOutput());
+        test(readableByteChannelInput(), writableByteChannelOutput());
+        test(readableByteChannelInput(), defaultOutput());
+    }
+
+    private static void test(InputStreamProvider inputStreamProvider, OutputStreamProvider outputStreamProvider)
+            throws Exception {
+        ifOutIsNullThenNpeIsThrown(inputStreamProvider);
+        contents(inputStreamProvider, outputStreamProvider);
+    }
+
+    private static void ifOutIsNullThenNpeIsThrown(InputStreamProvider inputStreamProvider) throws Exception {
+        try (InputStream in = inputStreamProvider.input()) {
+            assertThrowsNPE(() -> in.transferTo(null), "out");
+        }
+
+        try (InputStream in = inputStreamProvider.input((byte) 1)) {
+            assertThrowsNPE(() -> in.transferTo(null), "out");
+        }
+
+        try (InputStream in = inputStreamProvider.input((byte) 1, (byte) 2)) {
+            assertThrowsNPE(() -> in.transferTo(null), "out");
+        }
+    }
+
+    private static void contents(InputStreamProvider inputStreamProvider, OutputStreamProvider outputStreamProvider)
+            throws Exception {
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, new byte[0]);
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(1024, 4096));
+        // to span through several batches
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(16384, 16384));
```

Comment:
Should some random-sized buffers be tested? (Use `jdk.test.lib.RandomFactory` factory for this, not `j.u.Random`. The existing use of `Random` is fine.)

Should some testing be done of streams with non-zero initial position?

Reply:
Done in https://github.com/openjdk/jdk/pull/5209, will rebase on that PR once merged.